### PR TITLE
Yf sam locus and reference base iterator

### PR DESCRIPTION
**Pull request by yfarjoun**
This new small class holds a SamLocusIterator and a ReferenceFileWalker and keeps them in Tandem. This way you can easily know what the reference base is at the locus you are looking at.

### Checklist

- [ ] Code compiles correctly
- [ ] New tests covering changes and new functionality
- [ ] All tests passing
- [ ] Extended the README / documentation, if necessary
- [ ] Is not backward compatible (breaks binary or source compatibility)

### CODE REVIEW

File: src/main/java/htsjdk/samtools/reference/SamLocusAndReferenceIterator.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2009-2016 The Broad Institute
```
Comment (fleharty) [inline src/main/java/htsjdk/samtools/reference/SamLocusAndReferenceIterator.java]: 2018

File: src/main/java/htsjdk/samtools/reference/SamLocusAndReferenceIterator.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+import java.util.List;
+
+import static htsjdk.samtools.util.SamLocusIterator.*;
+
+/**
+ * Iterator that traverses a SAM File and a ReferenceFile, accumulating information on a per-locus basis.
+ * Only, whatever loci are covered by the input reads are returned.
+ * Duplicate reads and non-primary alignments are filtered out.
+ * Iterator element holds both pileup (in the form of a LocusInfo object) and the reference base
```
Comment (fleharty) [inline src/main/java/htsjdk/samtools/reference/SamLocusAndReferenceIterator.java]: end with "."

File: src/main/java/htsjdk/samtools/reference/SamLocusAndReferenceIterator.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import htsjdk.samtools.util.IterableOnceIterator;
+import htsjdk.samtools.util.SamLocusIterator;
+
+import java.util.List;
+
+import static htsjdk.samtools.util.SamLocusIterator.*;
+
+/**
+ * Iterator that traverses a SAM File and a ReferenceFile, accumulating information on a per-locus basis.
+ * Only, whatever loci are covered by the input reads are returned.
```
Comment (fleharty) [inline src/main/java/htsjdk/samtools/reference/SamLocusAndReferenceIterator.java]: "Only,"?
the comma seems strange

File: src/main/java/htsjdk/samtools/reference/SamLocusAndReferenceIterator.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        return locusIterator.hasNext();
+    }
+
+    @Override
+    public SAMLocusAndReference next() {
+        final LocusInfo locus = locusIterator.next();
+        final ReferenceSequence referenceSequence = referenceSequenceFileWalker.get(locus.getSequenceIndex(), locus.getSequenceName(),
+                locus.getSequenceLength());
+
+        //position is 1-based...arrays are 0-based!
```
Comment (fleharty) [inline src/main/java/htsjdk/samtools/reference/SamLocusAndReferenceIterator.java]: space after "//", capitalize P

File: src/main/java/htsjdk/samtools/reference/SamLocusAndReferenceIterator.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+ * Iterator element holds both pileup (in the form of a LocusInfo object) and the reference base
+ *
+ * @author Yossi Farjoun
+ */
+public class SamLocusAndReferenceIterator extends IterableOnceIterator<SamLocusAndReferenceIterator.SAMLocusAndReference> {
+
+    private final ReferenceSequenceFileWalker referenceSequenceFileWalker;
+    private final SamLocusIterator locusIterator;
+
+    public SamLocusAndReferenceIterator(final ReferenceSequenceFileWalker referenceFile, final SamLocusIterator locusIterator) {
```
Comment (fleharty) [inline src/main/java/htsjdk/samtools/reference/SamLocusAndReferenceIterator.java]: Should all these public methods have javadoc?

File: src/test/java/htsjdk/samtools/reference/SamLocusAndReferenceIteratorTest.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        overlapDetector.addAll(intervalList.getIntervals(), intervalList.getIntervals());
+
+        for (final SamLocusAndReferenceIterator.SAMLocusAndReference samLocusAndReference : samLocusAndReferences) {
+            Assert.assertEquals(samLocusAndReference.getRecordAndOffsets().size(), overlapDetector.overlapsAny(samLocusAndReference.locus) ? 2 : 0, "Position:" + samLocusAndReference.locus.toString());
+
+            for (final SamLocusIterator.RecordAndOffset recordAndOffset : samLocusAndReference.getRecordAndOffsets())
+                Assert.assertTrue(SequenceUtil.basesEqual(samLocusAndReference.referenceBase, recordAndOffset.getReadBase()), "Record: " + recordAndOffset.getRecord() + " Position:" + samLocusAndReference.locus.toString());
+        }
+    }
+}
```
Comment (fleharty) [inline src/test/java/htsjdk/samtools/reference/SamLocusAndReferenceIteratorTest.java]: new line to appease github?

File: src/test/java/htsjdk/samtools/reference/SamLocusAndReferenceIteratorTest.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import java.io.File;
+
+/**
+ * Created by farjoun on 6/1/18.
+ */
+public class SamLocusAndReferenceIteratorTest extends HtsjdkTest{
+    static final File TEST_DATA_DIR = new File("src/test/resources/htsjdk/samtools/reference");
```
Comment (fleharty) [inline src/test/java/htsjdk/samtools/reference/SamLocusAndReferenceIteratorTest.java]: this field could be private

File: src/test/java/htsjdk/samtools/reference/SamLocusAndReferenceIteratorTest.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        final SamLocusAndReferenceIterator samLocusAndReferences = new SamLocusAndReferenceIterator(referenceSequenceFileWalker, samLocusIterator);
+
+        IntervalList intervalList = new IntervalList(samReader.getFileHeader());
+        intervalList.add(new Interval("chrM", 1, 36));
+        intervalList.add(new Interval("chr20", 8401, 8460));
+
+        OverlapDetector<Interval> overlapDetector = new OverlapDetector<>(0, 0);
+        overlapDetector.addAll(intervalList.getIntervals(), intervalList.getIntervals());
+
+        for (final SamLocusAndReferenceIterator.SAMLocusAndReference samLocusAndReference : samLocusAndReferences) {
```
Comment (fleharty) [inline src/test/java/htsjdk/samtools/reference/SamLocusAndReferenceIteratorTest.java]: Could you provide a comment about what this is doing, it took me a while to figure it out.

File: src/main/java/htsjdk/samtools/reference/SamLocusAndReferenceIterator.java (hunk 5 above)
Comment (lbergelson) [inline src/main/java/htsjdk/samtools/reference/SamLocusAndReferenceIterator.java]: Should there be any sort of check that the reference is consistent with the sam file?

File: src/test/java/htsjdk/samtools/reference/SamLocusAndReferenceIteratorTest.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+import htsjdk.samtools.SamReader;
+import htsjdk.samtools.SamReaderFactory;
+import htsjdk.samtools.util.*;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import java.io.File;
+
+/**
+ * Created by farjoun on 6/1/18.
```
Comment (lbergelson) [inline src/test/java/htsjdk/samtools/reference/SamLocusAndReferenceIteratorTest.java]: I'm not a fan of these types of comments, git should track the provenance just fine.  That's very much just a personal preference though.

File: src/main/java/htsjdk/samtools/reference/SamLocusAndReferenceIterator.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    public SAMLocusAndReference next() {
+        final LocusInfo locus = locusIterator.next();
+        final ReferenceSequence referenceSequence = referenceSequenceFileWalker.get(locus.getSequenceIndex(), locus.getSequenceName(),
+                locus.getSequenceLength());
+
+        //position is 1-based...arrays are 0-based!
+        return new SAMLocusAndReference(locus, referenceSequence.getBases()[locus.getPosition() - 1]);
+    }
+
+    public static class SAMLocusAndReference {
```
Comment (lbergelson) [inline src/main/java/htsjdk/samtools/reference/SamLocusAndReferenceIterator.java]: I would make this final and add getters for the fields.

**Pull request merged**